Add a `path` field to `ValidationError`

gavadinov opened the pull request:
### Summary
On validation error make the path to the problematic field available to the client. Currently the path is only a field in `ValidationError` but later it can be used to improve the error messages (related to #100, #144)
### Example

##### Schema
```json
{
   "type":"object",
   "properties":{
      "foo":{
         "type":"object",
         "properties":{
            "bar":{
               "type":"integer"
            }
         }
      }
   }
}
```
##### Input
```json
{
  "foo": {
    "bar": "some string"
  }
}
```

##### Produced Error
`path: [foo, bar]`

Stranger6667 commented:
Awesome @gavadinov !

I will take a look at it tomorrow :)

picklenerd commented:
I made a separate PR for the public error stuff.

https://github.com/Stranger6667/jsonschema-rs/pull/185

Stranger6667 commented:
First of all, thank you for your efforts! I appreciate you working on this issue :)

I think that there should be an invariant - if any path stack is converted to JSON pointer, then looking up a value by this pointer should resolve into the validator's value. For example, let's take this schema:

```json
{
    "additionalItems": false,
    "items": [
        {
            "type": "string"
        }
    ]
}
```

If the additionalItems validator fails on `["a", 1]` , then the validation error should have `["additionalItems"]` as its path stack.

Python's `jsonschema` library behaves similarly:

```python
import jsonschema

try:
    jsonschema.validate(["a", 1], { "additionalItems": False , "items": [{"type": "string"}]})
except jsonschema.ValidationError as error:
    print(error.schema_path)

# Prints
# deque(['additionalItems'])
```

As far as I can see, it could be done via pushing needed keywords before (or just after) entering some validator compilation function.

**Tests**
Given the above invariant tests could be written like this:

```
#[test]
fn validator_path() {
    let validator_value = false;
    let schema = json!({ "additionalItems": validator_value , "items": [{"type": "string"}]});
    let value = json!(["a", 1]);
    let compiled = JSONSchema::compile(&schema).unwrap();
    let errors: Vec<ValidationError> = compiled.validate(&value).unwrap_err().collect();
    // Some routine to get a JSON pointer from the error schema path
    let pointer = format!("/{}", errors[0].path.as_ref().unwrap().join("/"));
    assert_eq!(schema.pointer(&pointer), Some(&json!(value)));
}
```
And then it could be parametrized with `test_case`. I'd appreciate some tests for common cases, including indexing within an array, so JSON pointers work. E.g., it will require `enumerate` in `ItemsArrayValidator::compile`

Here are some style nitpicks below

**Naming**
In the future, there could also be a path within the validated value. For this reason, I think that the path which is currently used in this PR could be named or derived from `schema_path`. In `CompilationContext` - `current_schema_path` and `schema_path` in all validation errors and validators.

**Arguments order**
I think that it will be nice to have this path information as the first argument whenever it is needed. E.g. it is the first argument to `ValidationError::additional_items` but the last one in compilation functions, e.g. in `AdditionalItemsBooleanValidator::compile`

gavadinov commented:
@Stranger6667 Thank you for the quick response. 

> if any path stack is converted to JSON pointer, then looking up a value by this pointer should resolve into the validator's value.

I'm not sure if this would fit the current implementation. The path that we generate now is not a path in the schema but a path in the JSON we are validating. Similar to the example here: #144 - this would let us know exactly which field in the JSON being validated failed validation.

For example the error produced by the example below would contain a path `[foo, bar]`:
##### Schema
```json
{
   "type":"object",
   "properties":{
      "foo":{
         "type":"object",
         "properties":{
            "bar":{
               "type":"integer"
            }
         }
      }
   }
}
```
##### Input
```json
{
  "foo": {
    "bar": "some string"
  }
}
```

We can add a new field that tracks the `schema_path` but that would have to touch more validators. Probably worth doing it separately?

Stranger6667 commented:
> I'm not sure if this would fit the current implementation. The path that we generate now is not a path in the schema but a path in the JSON we are validating. Similar to the example here: #144 - this would let us know exactly which field in the JSON being validated failed validation.

Aha, I completely missed that! Then, the JSON pointer invariant should work for the validated instance :) 

At the moment it seems like intermediate arrays within the schema lead to incomplete paths.

Schema:

```json
{
    "type": "array",
    "items": [
        {
            "type": "string"
        },
        {
            "type": "object",
            "properties": {
                "foo": {
                    "type": "integer"
                }
            }
        }
    ]
}
```

Value: 

```json
["a", {"foo": "bar"}]
```
The path should be `["1", "foo"]` but it is `["foo"]` now.

Also on the naming then - `current_instance_path` and `instance_path` might fit better than to distinguish it with schema paths.

> We can add a new field that tracks the schema_path but that would have to touch more validators. Probably worth doing it separately?

Absolutely!

gavadinov commented:
Ahh good catch on the array indices :see_no_evil: I pushed a new revision with:
* Unit Tests - I added them to `error.rs` but let me know if there is a better place for them.
* Keeping track of the path through arrays.
* Renamed `path` to `instance_path`/`current_instance_path`.

Stranger6667 commented:
Thanks!

However, I think that instance paths can't be unambiguously evaluated from schema only as the same validator can be used to validate different parts of the same input value. 

Schema:

```json
{
    "type": "array",
    "items": {
        "type": "string"
    }
}
```

Instance:

```json
["a", 42]
```

The current implementation returns `[]` as `instance_path`; however, it should be `["1"]`.

There is no information in the schema about how we can detect the particular failing index upfront. The `$ref` keyword also makes things more complicated, as any validator can be applied almost anywhere (including, e.g., using it from another arbitrary schema).

Therefore this information could only be obtained during validation and will have some runtime cost. But we can detect `schema_path` during the compilation phase.

gavadinov commented:
Ohh true. Is there a real use case for having the path be a valid JSON pointer? For my concrete use case I need to be able to obtain the name of the field that failed validation. So marking the arrays just with a `[]` would be fine: e.g.  `["foo", "bar", "[]", "baz"]`. Of course that is only for my use case, but I wonder if there is a real need for the concrete path that would justify the runtime cost :thinking:

Stranger6667 commented:
The original reason is that Draft 2019-09 [defines](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-02#section-10.3.1) output formats this way, but I think that anything else than a complete JSON pointer won't be unambiguous in some cases. For example, it won't be possible to say which element in a 2-dimensional array is invalid.

Schema:
```json
{
    "type": "array",
    "items": {
        "type": "array",
        "items": {
            "type": "integer"
        }
    }
}
```

Instance:
```json
[
    [1, 2, 3],
    [4, "5", 6], 
    [7, 8, 9]
]
```

And other cases of nested schemas. In other words, if the error reporting aims to show the precise position inside the validated instance, it won't be possible.

gavadinov commented:
@Stranger6667 I reimplemented everything from scratch and now the path is tracked at runtime. The tests from the first iteration still pass because the public interface is the same. I also added test cases for your examples above and they pass now :tada:

Stranger6667 commented:
This is amazing work! Thank you very much :) I will review it soon

Stranger6667 reviewed (approved):
I will merge it and then write some tests that will compare instance paths in errors from this library and Python's `jsonschema`. If they find something, then I'll fix them :) Then these changes will be released.

Thank you for your patience and efforts

gavadinov commented:
Perfect, thank you :partying_face:

Stranger6667 commented:
The change is released in `0.7.0` :tada:

Stranger6667 commented:
I forgot to implement a separate struct for `instance_path` to avoid some allocations. It is a breaking change and released in `0.8.0`. Sorry for the complications :(

This struct can be transformed to `Vec<String>` as it was before via the `.into_vec()` call:

```rust
error.instance_path.into_vec()
```

Or to JSON Pointer via `to_string()` or `Display` impl.

gavadinov commented:
That's amazing!! Thank you :tada:
